pymbolic/primitives.py

```diff
@@ -673,6 +694,38 @@ def __iter__(self):
         raise TypeError("expression types are not iterable")
 
 
+# {{{ attrs support
+
+def _attrs_init_arg_names(self) -> Sequence[str]:
+    from warnings import warn
+    warn("init_arg_names is deprecated and will be removed in 2023. "
+            "Use attrs.fields instead.",
+            DeprecationWarning, stacklevel=2)
+
+    return [fld.name for fld in attrs.fields(self.__class__)]
+
+
+def _attrs___getinitargs__(self) -> Sequence[Any]:
+    from warnings import warn
+    warn("__getinitargs__ is deprecated and will be removed in 2023. "
+            "Use attrs.fields instead.",
+            DeprecationWarning, stacklevel=2)
+
+    return tuple([getattr(self, fld.name)
+                    for fld in attrs.fields(self.__class__)])
+
+
+def augment_attrs_expression_class(cls) -> Type:
+    # FIXME Add 'if self is other' check to __eq__
```

Comment:
This is missing. We'll need this, or our hashing performance will be in the :toilet: .

Reply:
Implemented it.